source/extensions/filters/http/dynamic_forward_proxy/proxy_filter.cc

```diff
@@ -138,10 +147,51 @@ Http::FilterHeadersStatus ProxyFilter::decodeHeaders(Http::RequestHeaderMap& hea
   NOT_REACHED_GCOVR_EXCL_LINE;
 }
 
-void ProxyFilter::onLoadDnsCacheComplete(const Common::DynamicForwardProxy::DnsHostInfoSharedPtr&) {
-  ENVOY_STREAM_LOG(debug, "load DNS cache complete, continuing", *decoder_callbacks_);
+void ProxyFilter::addHostAddressToFilterState(
+    const Network::Address::InstanceConstSharedPtr& address) {
+
+  if (!config_->saveUpstreamAddress()) {
+    return;
+  }
+
+  if (!decoder_callbacks_ || !address) {
+    ENVOY_LOG_MISC(warn, "Bad parameter - decoder callbacks or address");
+    return;
+  }
+
+  ENVOY_STREAM_LOG(trace, "Adding resolved host {} to filter state", *decoder_callbacks_,
+                   address->asString());
+
+  const Envoy::StreamInfo::FilterStateSharedPtr& filter_state =
+      decoder_callbacks_->streamInfo().filterState();
+
+  if (!filter_state->hasData<StreamInfo::UpstreamAddressSet>(
+          StreamInfo::UpstreamAddressSet::key())) {
+    auto address_set = std::make_unique<StreamInfo::UpstreamAddressSet>();
+    address_set->addresses_.emplace(address);
+
+    filter_state->setData(StreamInfo::UpstreamAddressSet::key(), std::move(address_set),
+                          StreamInfo::FilterState::StateType::Mutable,
+                          StreamInfo::FilterState::LifeSpan::Request);
+
+  } else {
+    StreamInfo::UpstreamAddressSet& address_set =
+        filter_state->getDataMutable<StreamInfo::UpstreamAddressSet>(
+            StreamInfo::UpstreamAddressSet::key());
+    address_set.addresses_.clear();
```

Comment:
If you clear and then set what's the point of this? Just overwrite it above universally?